pandas/core/internals/construction.py

```diff
@@ -968,9 +972,31 @@ def convert(arr):
                     # i.e. maybe_convert_objects didn't convert
                     arr = maybe_infer_to_datetimelike(arr)
                     if dtype_backend != "numpy" and arr.dtype == np.dtype("O"):
```

Comment:
Based on some local testing using the ADBC driver I can confirm that it yields a `pandas.core.arrays.arrow.array.ArrowExtensionArray` with ` ._dtype` of `pandas.ArrowDtype`. When the query returns a bytes type column we get a `.type` of `bytes`, and likewise a `.type` of `string` is returned for a string type column. Seems like we don't need to do any conversions when using the ADBC driver as you've stated if I'm understanding correctly here! 

Wondering if it makes sense to remove the code here trying to convert based on a `dtype_backend != 'numpy'` since this will fix the cause of the exception in the issue? and maybe raise an exception when trying to use a `pyarrow` `dtype_backend` with the `SQLiteDatabase` connection type here: https://github.com/pandas-dev/pandas/blob/main/pandas/io/sql.py#L695 ?

Reply:
I think the general problem is that pandas does not have a first class "binary" data type, so I'm not sure how to solve this for anything but the pyarrow backend. 

With the pyarrow backend, I think you can still move this logic to `sql.py` and check the type of the column coming back from the database. If it is a binary type in the database, using the PyArrow binary type with that backend makes sense.

Not sure if @mroeschke has other thoughts to the general issue. This is likely another good use case to track in PDEP-13 https://github.com/pandas-dev/pandas/pull/58455

Reply:
I agree that this is the incorrect place to handle this conversion logic and this should only be a valid conversion for the pyarrow backend (`ArrowExtensionArray._from_sequence` should be able to return a binary type with binary data.)

Reply:
in `sql.py` it looks like the result of this conversion is being overwritten when the `dtype_backend` is `pyarrow`: https://github.com/pandas-dev/pandas/blob/main/pandas/io/sql.py#L181 and the dtype returned by the current logic is `ArrowDtype(pa.binary())` for the example in the issue, so maybe just removing the conversion logic is all that's needed to resolve this issue? I've removed the block doing the conversion and added a test case showing that the resulting df has a dtype of `ArrowDtype(pa.binary())` when the `dtype_backend='pyarrow'`